pydoctor/epydoc2stan.py

```diff
@@ -128,70 +140,74 @@ def look_for_intersphinx(self, name):
-             found, complain.
+        # Since there was no global match, go look for the name in the
+        # context where it was used.
 
-        """
+        # Check if fullID refers to an object by Python name resolution
+        # in our context. Walk up the object tree and see if fullID refers
+        # to an object by Python name resolution in each context.
         src = self.obj
         while src is not None:
             target = src.resolveName(fullID)
             if target is not None:
-                return self._objLink(target, prettyID)
+                return self._objLink(target)
             src = src.parent
-        target = self.obj.system.objForFullName(fullID)
-        if target is not None:
-            return self._objLink(target, prettyID)
-        fullerID = self.obj.expandName(fullID)
-        linktext = stdlib_doc_link_for_name(fullerID)
-        if linktext is not None:
-            return tags.a(tags.code(prettyID), href=linktext)
+
+        # Walk up the object tree again and see if fullID refers to an
+        # object in an "uncle" object.  (So if p.m1 has a class C, the
+        # docstring for p.m2 can say L{C} to refer to the class in m1).
+        # If at any level fullID refers to more than one object, complain.
         src = self.obj
         while src is not None:
             target = self.look_for_name(fullID, src.contents.values())
             if target is not None:
-                return self._objLink(target, prettyID)
+                return self._objLink(target)
             src = src.parent
+
+        # Examine every module and package in the system and see if fullID
+        # names an object in each one.  Again, if more than one object is
+        # found, complain.
         target = self.look_for_name(fullID, itertools.chain(
             self.obj.system.objectsOfType(model.Module),
             self.obj.system.objectsOfType(model.Package)))
         if target is not None:
-            return self._objLink(target, prettyID)
+            return self._objLink(target)
 
-        target = self.look_for_intersphinx(fullerID)
-        if not target:
-            # FIXME: https://github.com/twisted/pydoctor/issues/125
-            # expandName is unreliable so in the case fullerID fails, we
-            # try our luck with fullID.
-            target = self.look_for_intersphinx(fullID)
-        if target:
-            return tags.a(tags.code(prettyID), href=target)
-        if fullID != fullerID:
+        if fullID == fullerID:
+            self.obj.report(
+                "invalid ref to '%s' not resolved" % (fullID,),
+                section='resolve_identifier_xref')
+        else:
             self.obj.report(
                 "invalid ref to '%s' resolved as '%s'" % (fullID, fullerID),
-                section='translate_identifier_xref')
-        return tags.code(prettyID)
+                section='resolve_identifier_xref')
+        return None
```

Comment:
Yes, now that the presentation code is outside, an exception would make more sense. I'll have a look.

The motivation for this change was to reduce code duplication and to simplify the unit test, since the test can now just verify the URL instead of having to pick it out of an HTML or Stan fragment.